### ISSUE TITLE
Packages are downloaded insecurely

### ISSUE THREAD

**Issue by bos** [haskell/cabal#936]
(Imported from [Trac #946](http://hackage.haskell.org/trac/hackage/ticket/946), reported by cooldude on 2012-04-26)

It appears that when running cabal install package, the package is downloaded without any transport security.

Anyone who can perform a man in the middle attack could tamper with the package that is being downloaded, resulting in a complete compromise of the cabal user.

This makes it impossible to use cabal.

The servers should utilize TLS, it is possible to get a free certificate from startcom if price is a concern.

Additionally when packages are verified as non-malicious, they should be signed with a "cabal" signing key, and then the package signatures should be verified by cabal.

**Comment by alexanderkjeldaas** [comment]
I disagree with a lot of what is said in this discussion, and even the bug description!

@gregorycollins said: "Man in the middle attacks are much less of a problem than the other glaring security issues we have (lack of package signing and ability for anyone with a hackage account to push any package are much bigger ones.)".

This is misunderstanding the threat.  That type of attack on hackage is serious, but it will only happen once, and it will be fixed rapidly.  The whole eco-system will be upgraded after such an attack.  These kinds of attacks have happened on similar systems, for example on kernel.org, and the damage is minimal although the attack on the face of it, seems to be catastrophic.

It is similar to a deadly virus that is so deadly that the patient dies within 1 minute.  Such a virus can not spread, because it is too easy to detect, and therefore it is actually not that dangerous.

However, MITM attacks are silent.  They do not demand action.  We do not know how many people have already been compromised because of this bug, and the discussion on this issue shows the inherent problem of assessing MITM threats.

@23Skidoo  I also think including a dependency on OpenSSL in the platform when we have tls is a mistake.  Take any editor and open the source code for both, and it should be evident that one of these lives in the '80 with the security tradeoffs that were done at that time, and the other one is modern, high quality code.  Nobody should add OpenSSL as a dependency today if there is a good alternative.  This is not a controversial statement actually, just look at what Mozilla and Google decided wrt OpenSSL long before all the weaknesses were discovered.

But this should be moot, because TLS it not the solution to the underlying issue!

@bos regarding the issue description itself, transport security is not what is required.  There is no need for confidentiality when downloading packages.  What is needed is integrity, something that can be ensured by calculating SHA3.  This is 100 lines of code, as opposed to tls which is a lot bigger.  For example the cryptohash package could be added to the haskell platform.

This issue does not affect NixOs at all, since it always does checksum validation on downloaded source code.  This in itself proves that TLS is not a requirement.  NixOS must be attacked through its git repository, and git has an eco-system supporting signatures etc.  Nowhere in this chain of trust is TLS a requirement (though git depends on SHA1 which is weak).

Focusing on TLS is also wrong because it ignores the elephant in the room - namely that we trust the hackage server.  TLS assumes that the server is trusted, and this is something that we don't want, and as I have shown, don't need. This is completely unnecessary and the wrong focus.

So this is what I propose for stage1:
1. Adding SHA3 checksums to the `index.tar.gz` file.  This reduces the problem to securing `cabal update`, and paves the way for checking signatures.
2. Making cabal check SHA3 signatures for downloaded packages.
3. Adding `cryptohash`, `bytable`, and an `index.tar.gz` containing SHA3 checksums to the haskell platform.  This means that the haskell platform can boostrap any software from hackage securely, given that the haskell platform was downloaded securely.

To get signature support, either add the required libraries to the haskell platform, or make it so that the haskell platform includes a 'stage1' cabal that can boostrap a more featureful cabal from hackage.  The 'stage1' cabal can be the current cabal, only with the additional feature that it can use SHA3 to do integrity checks on downloaded packages.

I prefer bootstrapping, because it has the advantage that cabal is free to include all sorts of dependencies that are not in the haskell platform.  The only thing that needs to be in the haskell platform is SHA3 and the ability to check the integrity of downloaded packages.

The full cabal implementation can depend on `crypto-pubkey` which adds `crypto-random`, `crypto-numbers`, `crypto-pubkey-types`, `asn1-types`, `hourglass`, and `securemem`.

**Comment by dcoutts** [comment]
We're currently working on server-based index signing for the IHG.

**Comment by DaveCTurner** [comment]
Does this mean that this will be resolved by adding signatures to packages and not by supporting `https` URLs? That seems like the right thing to do, but in that case could #1616 be reopened as it's not a duplicate of this issue? Supporting `https` URLs solves a separate problem.

**Comment by DaveCTurner** [comment]
Additionally, I'm sure this has been discussed elsewhere, but just to be explicit:

- the integrity of the index file itself needs to be verified before the hashes it contains can be trusted.
- the system needs to protect against replay attacks too (as described at http://www.cs.arizona.edu/stork/packagemanagersecurity/ mentioned above)

For the first, GPG is a possible approach that avoids reimplementing too much crypto and infrastructure.

For the second, some kind of expiry mechanism would work, or perhaps just date the index (and sign the date) and let clients worry about how far back to trust things. GPG signatures include the date already.

**Comment by drbo** [comment]
What is the status of this?

**Comment by pikeas** [comment]
+1. New to Haskell, shocked that there's no mechanism for secure package download/install.

**Comment by drbo** [comment]
This makes cabal useless for me.  I don't have the time or skill to manually audit each package I download and am reduced to installing dependencies from source control, which is (fortunately) usually available over secure connection.

**Comment by 23Skidoo** [comment]
@drbo Check out http://www.well-typed.com/blog/2015/08/hackage-security-beta/

**Comment by 23Skidoo** [comment]
@drbo Also, the development version of `cabal-install` now uses `curl`/`wget` to download packages over HTTPS.

**Comment by drbo** [comment]
@23Skidoo That fixes it, thanks!

**Comment by phadej** [comment]
@23Skidoo what's the story on Windows?

**Comment by 23Skidoo** [comment]
@phadej On Windows it uses PowerShell. This feature could use more testing, BTW.

**Comment by magicgoose** [comment]
> this isn't a bug, it's a feature request

Of course this is a bug. The kind of bug which raises questions about the seriousness of Haskell community. Security (at least, against such trivial attacks) should not be an "additional feature", especially in such critical software component as the package manager.

But having this bug not fixed for 3 years — this is truly amazing, given how simple it is.

Looks like I should go to Homebrew and update the version, because
```
cabal-install version 1.22.6.0
using version 1.22.4.0 of the Cabal library
```
seems to be still using plaintext transfers, and the latest version is said to be better in this regard.

**Comment by mgsloan** [comment]
@magicgoose Agreed.  We do take security seriously in stack: https://www.fpcomplete.com/blog/2015/07/package-security-in-stack

**Comment by gbaz** [comment]
@mgsloan this is a bug tracker, not a place to discuss the merits of various systems. i'm removing this comment.

**Comment by mgsloan** [comment]
@gbaz OK. I just wanted him to know that the Haskell community does care about security.

**Comment by 23Skidoo** [comment]
I think that this can be marked as fixed given that 1.24 will have both HTTPS downloads and `hackage-security`.
